core_lib/src/activeframepool.cpp

```diff
@@ -87,12 +87,16 @@ void ActiveFramePool::onKeyFrameDestroy(KeyFrame* key)
     {
         mCacheFramesList.erase(it->second);
         mCacheFramesMap.erase(it);
+
+        // Just recalculate the total usage
+        // Not safe to call key->memoryUsage() here cuz it's in the KeyFrame's destructor
+        recalcuateTotalUsedMemory();
     }
 }
 
 void ActiveFramePool::discardLeastUsedFrames()
 {
-    while (mCacheFramesMap.size() > mMaxSize)
+    while (mTotalUsedMemory > mMemoryBudgetInBytes)
```

Comment:
I think there should also be a check here to stop if a minimum number of frames has been reached (probably matching the previous minimum for mMaxSize).

Consider the scenario where a single frame exceeds the memory budget. The frame would be immediately unloaded upon adding. I'll admit this scenario is the most unlikely given the 100 MiB minimum, but I've had images larger than that before, and it becomes much more likely if you consider that when scrubbing, `Object::updateActiveFrames` is called which loads all frames on all layers in a range that spans a total of 7 frames. On a project with three layers, each filled with frames, each scrub will load 21 frames, and some of those frames will be immediately unloaded if the average frame size exceeds ~4.7 MB, which is not all that outlandish for pngs.

At worst this causes serious issues for the code calling the loading which likely assumes that the image it has just called has been loaded. At best it causes unnecessary reloading of the files. Just skimming through the code, it looks as though paintImage will load the file on its own so there will be no visual issues, however since the frame is not the in the cache, it's resources will not be freed until it is added again, potentially keeping every frame in memory if you are scrubbing forward or playing under the right circumstances.

Reply:
Good point. Will improve it.